src/backend/cdb/cdbmutate.c

```diff
@@ -64,6 +64,19 @@
 
 #include "executor/executor.h"
 
+
+typedef struct ModifyTableMotionState
+{
+	Bitmapset  *resultRtis; 			/* Indexes into rtable for relations to
+										 * be modified */
+	bool		needExplicitMotion;
+	int			nMotionsAbove;			/* Number of Redistribute/Broadcast
```

Comment:
Why do we need int here? With the changes
```diff
diff --git a/src/backend/cdb/cdbmutate.c b/src/backend/cdb/cdbmutate.c
index e829fd3d1d5..eaf29c826a3 100644
--- a/src/backend/cdb/cdbmutate.c
+++ b/src/backend/cdb/cdbmutate.c
@@ -866,7 +866,7 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 
 			if (motion->motionType == MOTIONTYPE_HASH || motion->isBroadcast)
 			{
-				context->mt.nMotionsAbove += 1;
+				context->mt.nMotionsAbove = 1;
 			}
 		}
 
@@ -1154,7 +1154,7 @@ done:
 			/*
 			 * We're going out of this motion node.
 			 */
-			context->mt.nMotionsAbove -= 1;
+			context->mt.nMotionsAbove = 0;
 		}
 	}
```
all tests passed.

Reply:
```diff
--- results/DML_over_joins.out
+++ expected/DML_over_joins.out
@@ -1302,21 +1293,46 @@
 (4 rows)
 
 delete from s using r,sales where s.a = r.b and s.b = sales.id;
+ERROR:  distribution key of the tuple (0, 2) doesn't belong to current segment (actually from seg0) (nodeModifyTable.c:716)
+explain (costs off) delete from s using r,sales where s.a = r.b and s.b = sales.id;
+QUERY PLAN
+___________
+ Delete on s
+   ->  Hash Join
+         Hash Cond: (r.b = s.a)
+         ->  Seq Scan on r
+         ->  Hash
+               ->  Broadcast Motion 3:3  (slice2; segments: 3)
+                     ->  Hash Join
+                           Hash Cond: (sales.id = s.b)
+                           ->  Broadcast Motion 3:3  (slice1; segments: 3)
+                                 ->  Seq Scan on sales
+                           ->  Hash
+                                 ->  Seq Scan on s
+GP_IGNORE:(13 rows)
```

Reply:
May be some words about in description?